Feature/lock

a981008 opened the pull request:
refs: https://github.com/nacos-group/r-nacos/issues/243

heqingpan committed:
feat: 初步完成raft对外接入重构，把与外部的数据交互汇总在raftdata::RaftDataHandler中

heqingpan committed:
chore: 调整actions run_tests触发的范围

heqingpan committed:
feat: 完成在raft中增加sequence表，管理向上增长的id; #241

heqingpan committed:
fmt: cargo fmt & cargo fix

heqingpan committed:
feat(sequence): 添加序列相关接口并更新 AppShareData 结构

- 在 AppShareData 结构中添加 SequenceDbManager 和 SequenceManager
- 新增序列调试接口 /rnacos/debug/sequence/next_id
- 实现获取下一个序列 ID 的功能
- 优化命名服务日志输出

heqingpan committed:
feat: 初步完成McpManager基本结构开发

heqingpan committed:
feat: 调整mcp tools的持久化结构; #241

a981008 committed:
feat: 分布式锁实现

a981008 committed:
feat: 分布式锁实现

a981008 committed:
feat: 分布式锁使用示例

a981008 committed:
feat: 分布式锁使用示例

heqingpan commented:
本周末人都在外面，晚点处理。

在 #241 mcp相关支持的分支中对raft存储有做一些代码重构应该会有些冲突，到时处理估计要花些时间。

a981008 committed:
refactor: 在释放时删除失效的锁。

a981008 committed:
refactor: 在释放时删除失效的锁。

a981008 committed:
Merge branch 'nacos-group:master' into feature/lock

heqingpan committed:
feat: 初步完成Mcp ToolSpec的核心功能开发; #241

heqingpan committed:
feat: 初步完成Mcp Server核心功能开发; #241

heqingpan commented:
还没有运行代码验证，初步看了一下代码，把看到的问题点列一下。

1. LockActor中的表数据接入raft，不过只接入log，没有接入snapshot打包和加载。

raft为了避免log无限增长，每写入一定量（默认是1万，可配置）log,就会触发snapshot。触发snapshot时会把当前数据全部写入snapshot文件，然后标记snapshot后的日志点位；下次重启时会先加载snapshot，再加载snapshot点位后的日志。

如果一个表数据没有接入snapshot,那么在snapshot后重启之后，或者snapshot后有新节点接入就会数据丢失问题。

2. T_LOCK表数据没有接入数据备份与恢复，备份数据恢复后锁信息会丢失。

3. lock expire操作暂还不支持。

heqingpan commented:
可能参考`NamespaceActor`,实现`impl Handler<RaftApplyDataRequest> for NamespaceActor`,同时在调用点加上对LockActor的调用完成对raft snapshot的接入。

heqingpan commented:
关于raft 锁的支持，我之前有设想是可以和redis类似通过自带的缓存(`CacheManager`)来实现。

在ratch-job中，实现已经实现了类型redis锁的功能（ [ratch-job cache_core](https://github.com/ratch-job/ratch-job/blob/master/src/cache/core.rs) ），但r-nacos的`CacheManager`目前不是直接接入raft,而是通过`TableManager`，还不能直接完备支持锁的功能。

cache中支持缓存锁相关代码片段：

```rust
impl Handler<CacheManagerRaftReq> for CacheManager {
    type Result = anyhow::Result<CacheManagerRaftResult>;

    fn handle(&mut self, req: CacheManagerRaftReq, _ctx: &mut Self::Context) -> Self::Result {
        match req {
            CacheManagerRaftReq::Set(set_info) => Ok(self.set(set_info)),
            CacheManagerRaftReq::GetSet(set_info) => Ok(self.get_set(set_info)),
            CacheManagerRaftReq::Get(key) => Ok(self.get_value(&key)),
            CacheManagerRaftReq::Remove(key) => Ok(self.remove(&key)),
            CacheManagerRaftReq::Exists(key) => Ok(self.exists(&key)),
            CacheManagerRaftReq::Expire(key, expire) => Ok(self.expire(key, expire)),
            CacheManagerRaftReq::Ttl(key) => Ok(self.get_ttl(&key)),
            CacheManagerRaftReq::Incr(key, expire) => Ok(self.incr(key, expire)),
            CacheManagerRaftReq::Decr(key, expire) => Ok(self.decr(key, expire)),
        }
    }
}
```

之前提到接入mcp时对r-naocs raft有些重构，它的目标之一也是包括给现有的cache提供更强的能力。

---
上面是我之前对锁支持方式的思考。

本次我们可以先按你这次加的LockActor实现锁的功能，后面时机合适再考虑是否切换到cache支持锁。

heqingpan committed:
feat: 初步完成mcp_tool_spec_api控制台接口开发；241

heqingpan committed:
feat: 增加mcp工具toolspec控制台接口的集成测试用例; #241

heqingpan committed:
fix: 1.调整add_or_update_tool_spec接口，每次更新时都会设置一个新的toolspec version; #241
2. 调整toolspec_console_api_test.py，使其接口集成测试能整体通过

heqingpan committed:
feat: update actor_model ToolSpecDto

heqingpan committed:
refactor(toolspec): 更新 ToolSpec 结构并验证 'function' 字段

- 将 ToolSpec 中的 'parameters' 字段重命名为 'function'
- 更新相关模型和 API 调用以适应新结构
- 增加对 'function' 字段格式的验证
- 调整测试用例以符合新的数据结构

heqingpan committed:
refactor(mcp): 重构工具规格引用计数逻辑

-移除 ToolSpecVersion 中的 ref_count 字段
- 在 McpManager 中添加 tool_spec_version_ref_map 用于记录引用计数
- 更新工具规格时，计算引用计数的增减差值并更新到 tool_spec_version_ref_map
- 负数的引用计数现在表示减少引用
- 优化了工具规格引用计数的管理和存储逻辑

heqingpan committed:
feat: 初步完成mcp server控制台管理接口开发; #241

heqingpan committed:
feat(console): 添加 McpServer 控制台接口路由

- 新增 McpServer 列表查询、详情获取、添加、更新、删除等接口
- 实现 McpServer 历史记录查询、当前版本发布、历史版本发布等功能
#241

heqingpan committed:
test: 更新mcp_server集成注册用例; #241

heqingpan committed:
feat: 调整tools的ToolFunctionValue.parameters 类型定义; #241

heqingpan committed:
feat: 更新McpServerDto字段定义

heqingpan committed:
refactor(mcp): 优化 McpServerDto 新增 release_value处理逻辑

- 修改 public_history 函数，将 current_value 更改为 release_value
- 优化 McpServerDto 构造函数，增加对 release_value 的处理
- 删除 McpServerDto 中冗余的代码
#241

heqingpan committed:
feat(sdk-examples): 添加 Python 计算服务 API 示例

- 新增 calculate_api 项目，实现基本数学运算的 FastAPI 服务
- 集成 Nacos SDK，自动注册服务到 r-nacos 服务注册中心
- 提供详细的 README 文档，说明依赖安装、运行方法和 API 接口
- 添加 .gitignore 和 pyproject.toml 文件，配置项目环境

heqingpan committed:
refactor(sdk-examples):重构 Nacos 服务注册逻辑- 将 Nacos 服务器地址和命名空间配置移至全局变量
- 优化 register_to_nacos 函数，增加 client 参数
- 调整服务实例信息配置
- 重新组织代码结构，提高可读性和可维护性

heqingpan committed:
feat(openapi): 添加 MCP 接口支持

- 新增 MCP API 处理逻辑，包括初始化、工具调用等功能
- 添加 MCP 相关的模型定义和序列化结构
- 在 Web 配置中注册 MCP 接口路由
- 优化工具函数包装结构，支持更丰富的工具信息展示
- mcp http-streamable协议中除为"tools/call"未适配，其它的接口都已初步完成。
#241

heqingpan committed:
feat(mcp): 实现mcp tools/call 接口并优化相关功能

- 新增 handle_tools_call函数，实现工具调用逻辑- 添加 select_tool_and_url 函数，用于选择工具和构建 URL- 在 AppShareData 中添加 common_client 字段，用于 HTTP 请求- 实现 ToolRouteRule 中的 is_need_host 和 build_url 方法，优化 URL 处理
- 更新 NamingCmd 和 NamingResult 枚举，支持 SelectOneInstance 功能- 修改 Service 实现 select_one_instance 方法，用于随机选择实例
#241

heqingpan committed:
feat: metrics接口补充content-type信息，以适配最新版本的prometheus；#247

heqingpan committed:
upgrade version to v0.6.22

heqingpan committed:
refactor(mcp): 重构命名空间过滤逻辑

- 将 McpQueryParam 中的 namespace_filter 字段重命名为 namespace_id
- 修改 namespace_id 的处理逻辑，从模糊匹配改为精确匹配

heqingpan committed:
refactor(mcp): 更新工具版本初始化逻辑

- 将工具版本初始化为参数中的 version 值，而不是固定为 1
- 在版本映射中使用参数中的 version 值作为键，而不是固定为 1

heqingpan committed:
refactor(mcp): 重构工具函数相关代码

- 移除了 ToolFunctionWrap 结构体，直接使用 ToolFunctionValue
- 更新了 tools/list 方法的处理逻辑
- 调整了 ToolFunctionValue 结构体的命名
#241

heqingpan committed:
feat(mcp): 添加数据迁移功能

- 在 McpManager 中实现 transfer_backup 方法，用于备份MCP工具和服务器信息
- 在 McpServer 中添加 unique_key 字段，用于唯一标识服务器
- 更新数据对象定义，添加 unique_key 字段- 修改数据迁移模块，支持 MCP 数据备份
#241

heqingpan committed:
feat(transfer): 支持 MCP 数据导入

- 新增 MCP服务器和工具规格导入功能
- 实现 MCP 相关的 TransferReader 和 TransferWriter 支持- 添加 McpImportContext 用于重置 MCP 数据 ID
- 在 McpManager 中添加处理 MCP导入请求的方法
- 更新 TransferImportManager 以支持 MCP 导入
#241

heqingpan committed:
refactor(mcp): 移除工具版本映射中的 ID 字段

- 从 McpToolDo、McpSimpleTool 和 McpTool 结构体中移除了 ID 字段
- 更新了工具版本映射逻辑，不再使用 ID进行映射
- 优化了工具版本获取逻辑，当指定版本不存在时使用当前版本
#241

heqingpan committed:
feat(mcp): 优化 MCP服务发现机制并调整 API路由

- 新增 GetServerByKey 请求类型，用于通过服务器唯一键获取服务器信息
- 更新 API 路由，使用 server_key 和 auth_key 替代原有的 id 和 key
- 在 McpServerParam 中添加 unique_key 和 publish_value_id 字段
- 修改服务器添加和更新逻辑，支持唯一键和发布版本 ID
- 调整MCP OpenAPI 路径参数，使用 server_key 和 auth_key
#241

heqingpan committed:
fix(mcp): 修复删除工具规格时的错误处理

- 在删除工具规格前，检查该规格是否被其他版本引用
- 如果工具规格被使用，则返回错误，提示"tool spec is used"
- 修改 McpManagerRaftReq::RemoveToolSpec 处理逻辑，添加错误处理
#241

heqingpan committed:
feat(mcp): 新增MCP SSE (Server-Sent Events) 支持

- 添加 SSE 连接管理功能，实现心跳检测和消息发送
- 新增 SSE 相关的 API 接口和处理逻辑
- 更新 AppShareData 结构，增加 SseStreamManager
- 重构部分代码以支持 SSE功能
#241

heqingpan committed:
feat(mcp): mcp sse请求支持在r-nacos集群不同节点调用 #241

heqingpan committed:
feat(user): 新增 MCP 工具规格和服务器管理权限

- 添加了 MCP 工具规格访问和管理权限
- 添加了 MCP 服务器访问和管理权限
- 将新权限集成到现有的访问控制体系中
#241

heqingpan committed:
feat(toolbox): 添加工具规格批量更新功能

- 在 ActorModel 中新增 UpdateToolSpecList 消息类型
- 在 Console API 中添加批量更新工具规格的路由
- 在 McpCore 中实现批量更新工具规格的逻辑
- 在 MCP Tool Spec API 中添加批量更新工具规格的处理函数

heqingpan committed:
feat(mcp): 增强 MCP 工具功能和权限管理

- 修复 McpServerQueryRequest 中 namespace_id 的空值处理逻辑
- 增强 ToolRouteRule 的 URL 构建功能，支持绝对路径和相对路径
- 新增 ConvertType 枚举，支持 JsonToForm 和 JsonToUrl 转换类型
- 优化工具调用请求处理，支持不同的内容类型和参数转换方式
- 添加 MCP 工具规格批量更新权限配置

heqingpan committed:
refactor(mcp): 优化 add_mcp_server 接口的错误处理机制

- 重构 add_mcp_server 函数，拆分为 add_mcp_server 和 do_add_mcp_server
- 使用 anyhow::Result 和 ? 操作符简化错误处理逻辑
- 新增统一的 handle_error 函数处理 anyhow::Error 类型错误
- 减少繁琐的嵌套 match 语句，提高代码可读性和维护性
- 保持原有功能完整性的同时优化代码结构

heqingpan committed:
style: 代码格式化优化

- 移除未使用的 std::ops::Index 导入
- 优化代码格式和空格布局
- 提升代码可读性

heqingpan committed:
refactor(mcp): 重构 MCP 服务器 API 的错误处理机制

提取 update_mcp_server 和 remove_mcp_server 函数的核心逻辑到独立函数，简化主函数结构，提高代码可读性和可维护性。

heqingpan committed:
refactor(mcp): 简化MCP服务器API的错误处理机制

- 移除未使用的导入和冗余的调试日志
- 重构publish_current_mcp_server和publish_history_mcp_server函数
- 统一错误处理模式，使用更简洁的?操作符
- 简化响应处理逻辑，提高代码可读性

heqingpan committed:
refactor(mcp): 简化MCP工具规格API的日志记录和代码逻辑

- 移除冗余的debug和info级别日志记录
- 简化成功响应的处理逻辑
- 清理未使用的参数和变量
- 提高代码可读性和简洁性

heqingpan committed:
fix(mcp): 修复MCP服务器ID重复分配问题

在设置MCP服务器时，如果服务器已存在，则保留原有ID而不是重新分配，避免ID冲突。

heqingpan committed:
feat(mcp): 添加ToolSpec批量导出功能 #241

heqingpan committed:
feat(mcp): 添加ToolSpec批量导入功能，支持通过ZIP文件导入多个工具规格 #241

heqingpan committed:
feat(calculate-api): 增强Python计算API支持多种请求格式 #241

heqingpan committed:
Merge remote-tracking branch 'origin/master' into develop

heqingpan committed:
upgrade version to v0.7.0-alpha

heqingpan committed:
优化reqwest依赖配置，禁用默认特性并添加json支持

heqingpan committed:
feat(calculate-api): 添加Rust计算服务API示例

新增基于Actix Web的计算服务，支持加、减、乘、除运算，
提供JSON、Form表单和Query参数三种请求格式，
集成Nacos服务注册发现功能，用于验证r-nacos MCP服务转发。

heqingpan committed:
feat(mcp): 增强MCP API日志记录和错误处理 #241

heqingpan committed:
upgrade version to v0.7.0-beta

heqingpan committed:
fix: 修复MCP ToolSpec描述字段错误并更新版本到v0.7.0-beta2

heqingpan committed:
upgrade version to v0.7.0

heqingpan commented:
支持mcp服务已发布，这个功能计划近期处理。

其中的raft存储层，可能会直接换成r-nacos现有的缓存（其缓存核心功能还需要先优化补充）。

heqingpan committed:
docs: 更新README文档并添加MCP相关图片

heqingpan committed:
feat(mcp): 添加prompts/list和ping方法支持并优化SSE兼容性

heqingpan committed:
upgrade version to v0.7.1

a981008 committed:
Merge branch 'feature/lock' of github.com:a981008/r-nacos into feature/lock

# Conflicts:
#	src/common/constant.rs
#	src/lib.rs
#	src/raft/cluster/model.rs
#	src/raft/cluster/route.rs
#	src/raft/filestore/raftapply.rs
#	src/raft/filestore/raftdata.rs
#	src/raft/store/mod.rs
#	src/starter.rs
